devmapper: Drop devices lock before returning from function

rhvgoyal opened the pull request:
cleanupDeleted() takes devices.Lock() but does not drop it if there are
no deleted devices. Hence docker deadlocks if one is using deferred
device deletion feature. (--storage-opt dm.use_deferred_deletion=true).

Fix it. Drop the lock before returning.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

rhvgoyal commented:
CC @LK4D4 @vbatts @calavera 

This single line fix will need to go in 1.9. PTAL.

cpuguy83 commented:
unit test?

LK4D4 commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -599,6 +599,7 @@ func (devices *DeviceSet) cleanupDeletedDevices() error {
 
 	// If there are no deleted devices, there is nothing to do.
 	if devices.nrDeletedDevices == 0 {
+		devices.Unlock()
```
let's do `defer`, so next time when someone decide to fix this function won't forget to add unlock before return.

rhvgoyal replied:
@LK4D4 

Are you saying do "defer devices.Unlock()" right after taking lock? I can't do that and reason being that with-in same function I need to drop the lock before calling devices.DeleteDevice(). And DeleteDevice() assumes that it will be called without devices.Lock() held.

LK4D4 replied:
Ok, but it's definitely something wrong with locks in devicemapper. It shouldn't be like this.

rhvgoyal replied:
@lk4d4 I agree that this looks odd. In general reason is that when device deletion is happning, we can try deletion/deactivation in a loop. And that that time we don't want to block rest of the device operations in parallel. So we drop the inner devices lock while continue to hold per device lock. But that also seems to require that per device lock be acquired before devices lock and all these complications arise. 

I hope there is a better way to design this. I will keep this in mind.

rhvgoyal commented:
@cpuguy83 Right now there are unit tests for deferred device deletion feature. I can look into adding some though.

cpuguy83 commented:
@rhvgoyal In this case, a test just on this function would suffice, but also +1 to @LK4D4 there are some greater locking issues here.

rhvgoyal commented:
@cpuguy83 ok, I am trying to write a unit test. Is there a way to figure out if lock is currently held or not. Currently I am taking the lock after the call to the function and if lock has been released test will pass otherwise it will hang. But hanging will stop execution of other tests too.

So there has to be a better way to do it. Last time I looked, I could not find a way to check if a lock is locked or not without blocking on the lock.

LK4D4 commented:
@rhvgoyal spawn in goroutine and wait for some amount of time in select on some chan.

rhvgoyal commented:
@LK4D4 sounds good. Will do.

rhvgoyal commented:
@cpuguy83 @LK4D4 I added a unit test case for locking issue. PTAL.

LK4D4 commented on daemon/graphdriver/devmapper/devmapper_test.go:
```diff
@@ -79,3 +80,36 @@ func testChangeLoopBackSize(t *testing.T, delta, expectDataSize, expectMetaDataS
+	go testTakeReleaseLock(driver.DeviceSet, doneChan)
+
+	select {
+	case <-timer.C:
```
This can be just `time.After(...)`

LK4D4 replied:
Also, I think you can safely omit `Unlock` here to simplify test.

rhvgoyal replied:
@lk4d4 Will use time.After()

If timer expires, then I have to do unlock(). Otherwise when I call into the driver to Shutdown(), it hangs and whole unit test machinery hangs.

LK4D4 replied:
Ah, okay.

cpuguy83 commented on daemon/graphdriver/devmapper/devmapper_test.go:
```diff
@@ -79,3 +80,36 @@ func testChangeLoopBackSize(t *testing.T, delta, expectDataSize, expectMetaDataS
+	}
+}
+
+func testTakeReleaseLock(devices *DeviceSet, doneChan chan<- bool) {
```
Can just define this above as "go func() ..."

rhvgoyal replied:
ok, will do.

rhvgoyal committed:
devmapper: Drop devices lock before returning from function

cleanupDeleted() takes devices.Lock() but does not drop it if there are
no deleted devices. Hence docker deadlocks if one is using deferred
device deletion feature. (--storage-opt dm.use_deferred_deletion=true).

Fix it. Drop the lock before returning.

Also added a unit test case to make sure in future this can be easily
detected if somebody changes the function.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@cpuguy83 @LK4D4 Took care of review comments. PTAL.

LK4D4 commented:
LGTM, thanks!

rhvgoyal commented:
@vbatts @cpuguy83 PTAL. I am kind of stuck because of this deferred device deletion is not working at all and till then I can't enable it in fedora.

cpuguy83 commented:
Sorry, was verifying test fails on master.
LGTM

cpuguy83 commented:
ping @tiborvass for 1.9

rhvgoyal commented:
@tiborvass ping. Can you pull this fix in 1.9.
